pkg/segment/writer/agiletree.go

```diff
@@ -279,6 +280,149 @@ func (stb *StarTreeBuilder) insertIntoTree(node *Node, colVals []uint32, recNum
 	}
 }
 
+func (stb *StarTreeBuilder) updateAggVals(node *Node, nodeToMerge *Node) error {
+	if nodeToMerge.aggValues == nil {
+		return nil
+	}
+	if node.aggValues == nil {
+		node.aggValues = make([]utils.CValueEnclosure, len(stb.mColNames)*TotalMeasFns)
+	}
+
+	var err error
+	for mcNum := range stb.mColNames {
+		midx := mcNum * TotalMeasFns
+		agvidx := midx + MeasFnMinIdx
+		node.aggValues[agvidx], err = utils.Reduce(node.aggValues[agvidx], nodeToMerge.aggValues[agvidx], utils.Min)
+		if err != nil {
+			log.Errorf("updateAggVals: error in min err:%v", err)
+			return err
+		}
+
+		agvidx = midx + MeasFnMaxIdx
+		node.aggValues[agvidx], err = utils.Reduce(node.aggValues[agvidx], nodeToMerge.aggValues[agvidx], utils.Max)
+		if err != nil {
+			log.Errorf("updateAggVals: error in max err:%v", err)
+			return err
+		}
+
+		agvidx = midx + MeasFnSumIdx
+		node.aggValues[agvidx], err = utils.Reduce(node.aggValues[agvidx], nodeToMerge.aggValues[agvidx], utils.Sum)
+		if err != nil {
+			log.Errorf("updateAggVals: error in sum err:%v", err)
+			return err
+		}
+
+		agvidx = midx + MeasFnCountIdx
+		node.aggValues[agvidx], err = utils.Reduce(node.aggValues[agvidx], nodeToMerge.aggValues[agvidx], utils.Count)
+		if err != nil {
+			log.Errorf("updateAggVals: error in count err:%v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (stb *StarTreeBuilder) cleanupCommonChildren(currNode *Node) error {
+	var err error
+	// delink the children from the current nodes and accumulate them as commonChildren
+	for _, nodes := range currNode.commonChildren {
+		fixedNode := nodes[0]
+		commonChildren := make(map[uint32][]*Node)
+		for _, node := range nodes {
+			for key, child := range node.children {
+				if commonChildren[key] == nil {
+					commonChildren[key] = []*Node{child}
+				} else {
+					commonChildren[key] = append(commonChildren[key], child)
+					err = stb.updateAggVals(commonChildren[key][0], child)
+					if err != nil {
+						return err
+					}
+				}
+				child.parent = fixedNode
+				delete(node.children, key)
+			}
+		}
+		fixedNode.commonChildren = commonChildren
+	}
+
+	// link parent and children properly and cleanup commonChildren
+	for _, children := range currNode.commonChildren {
+		err = stb.cleanupCommonChildren(children[0])
+		if err != nil {
+			return err
+		}
+		currNode.children[children[0].myKey] = children[0]
+	}
+	currNode.commonChildren = nil
+
+	return nil
+}
+
+func (stb *StarTreeBuilder) updateLastLevel(node *Node) error {
+	for _, child := range node.children {
+		err := stb.updateAggVals(node, child)
+		if err != nil {
+			return err
+		}
+		delete(node.children, child.myKey)
+	}
+
+	return nil
+}
+
+func (stb *StarTreeBuilder) removeLevelFromTree(node *Node, currIdx uint, idxToRemove uint, lastIdx uint) error {
```

Comment:
nit: at some point it would be good to make a wrapper for this like `removeColumn(colName string)` that just calls this method with the correct parameters. That would also make your unit tests a little more clear because you can test on `removeColumn()` instead of testing `removeLevelFromTree()` directly

Reply:
addressed in the latest commit.